### ISSUE TITLE
vmware.vmware_rest collection fails on ansible tower

### ISSUE THREAD

**Issue by RobVerduijn** [ansible-collections/vmware.vmware_rest#168]
##### SUMMARY
When using the wmware.vmware_rest collection on ansible tower it fails while complaining about missing python modules.

In my test case it complains about not being able to find aoihttp.
I have checked Ansible Tower, and the ansible tower installation does contain this python module.

For some reason the collection can't seem to find that python module.

This all works fine when using ansible on the command line from a centos8 controller

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

the vmware.vmware_rest collection used on ansible tower
##### ANSIBLE VERSION

```
ansible 2.9.15
ansible tower 3.7.3

```

##### CONFIGURATION

```
```

##### STEPS TO REPRODUCE

create a git project containing the play and an inventory
in tower:
configure a secret if you are using vaults in your project
set up a project in tower using the git as a source
set up an inventory using the project as a source (the project must contain an inventory )
set up a template using the previous items
launch a job from the template.

```- name: issue reproducer
  hosts: localhost
  become: false
  gather_facts: false
  vars:
    vcenter_host: vsphere_host
    vcenter_username: admin
    vcenter_password: secret
  collections:
  - vmware.vmware_rest

  tasks:
  - name: get vm moid
    vmware.vmware_rest.vcenter_vm_info:
      vcenter_hostname: "{{ vcenter_host }}"
      vcenter_password: "{{ vcenter_password }}"
      vcenter_username: "{{ vcenter_username }}"
      vcenter_validate_certs: false
      filter_names: centos8
    register: vm_find_res
    until: not vm_find_res['failed']
    retries: 720
    delay: 5
```

##### EXPECTED RESULTS

the play to finish

##### ACTUAL RESULTS

```
{
    "msg": [
        "  File \"/usr/lib64/python3.6/runpy.py\", line 193, in _run_module_as_main\n    \"__main__\", mod_spec)\n",
        "  File \"/usr/lib64/python3.6/runpy.py\", line 85, in _run_code\n    exec(code, run_globals)\n",
        "  File \"/tmp/ansible_vmware.vmware_rest.vcenter_vm_info_payload_iuig08u9/ansible_vmware.vmware_rest.vcenter_vm_info_payload.zip/ansible_collections/cloud/common/plugins/module_utils/turbo/server.py\", line 231, in <module>\n",
        "  File \"/tmp/ansible_vmware.vmware_rest.vcenter_vm_info_payload_iuig08u9/ansible_vmware.vmware_rest.vcenter_vm_info_payload.zip/ansible_collections/cloud/common/plugins/module_utils/turbo/server.py\", line 210, in start\n",
        "  File \"/usr/lib64/python3.6/asyncio/base_events.py\", line 438, in run_forever\n    self._run_once()\n",
        "  File \"/usr/lib64/python3.6/asyncio/base_events.py\", line 1451, in _run_once\n    handle._run()\n",
        "  File \"/usr/lib64/python3.6/asyncio/events.py\", line 145, in _run\n    self._callback(*self._args)\n",
        "  File \"/tmp/ansible_vmware.vmware_rest.vcenter_vm_info_payload_iuig08u9/ansible_vmware.vmware_rest.vcenter_vm_info_payload.zip/ansible_collections/cloud/common/plugins/module_utils/turbo/server.py\", line 192, in handle\n",
        "Traceback (most recent call last):\n  File \"/tmp/ansible_vmware.vmware_rest.vcenter_vm_info_payload_iuig08u9/ansible_vmware.vmware_rest.vcenter_vm_info_payload.zip/ansible_collections/cloud/common/plugins/module_utils/turbo/server.py\", line 150, in run\n  File \"/tmp/ansible_vmware.vmware_rest.vcenter_vm_info_payload_iuig08u9/ansible_vmware.vmware_rest.vcenter_vm_info_payload.zip/ansible_collections/vmware/vmware_rest/plugins/modules/vcenter_vm_info.py\", line 378, in main\n    log_file=module.params[\"vcenter_rest_log_file\"],\n  File \"/tmp/ansible_vmware.vmware_rest.vcenter_vm_info_payload_iuig08u9/ansible_vmware.vmware_rest.vcenter_vm_info_payload.zip/ansible_collections/vmware/vmware_rest/plugins/module_utils/vmware_rest.py\", line 24, in open_session\n    aiohttp = importlib.import_module(\"aiohttp\")\n  File \"/usr/lib64/python3.6/importlib/__init__.py\", line 126, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n  File \"<frozen importlib._bootstrap>\", line 994, in _gcd_import\n  File \"<frozen importlib._bootstrap>\", line 971, in _find_and_load\n  File \"<frozen importlib._bootstrap>\", line 953, in _find_and_load_unlocked\nModuleNotFoundError: No module named 'aiohttp'\n"
    ],
    "invocation": {
        "module_args": {
            "vcenter_hostname": "vsphere_host",
            "vcenter_password": "VALUE_SPECIFIED_IN_NO_LOG_PARAMETER",
            "vcenter_username": "admin",
            "vcenter_validate_certs": false,
            "filter_names": [
                "centos8"
            ],
            "vcenter_rest_log_file": null,
            "filter_clusters": null,
            "filter_datacenters": null,
            "filter_folders": null,
            "filter_hosts": null,
            "filter_power_states": null,
            "filter_resource_pools": null,
            "filter_vms": null,
            "vm": null
        }
    },
    "_ansible_no_log": false,
    "attempts": 1,
    "changed": false,
    "retries": 721
}

```

**Comment by goneri** [comment]
Thank you for your report @RobVerduijn.

**Comment by goneri** [comment]
According to the error message above, Ansible uses Python 3.6 to run the module and aiohttp is indeed not available (`No module named 'aiohttp'`).
What make you think aiohttp is actually available?

**Comment by RobVerduijn** [comment]
I looked on the system and saw that it was present, it is part of the tower package.

Rob

**Comment by RobVerduijn** [comment]
Does this answer your question ?

**Comment by goneri** [comment]
Hi @RobVerduijn, Not really because Python cannot find `aiohttp`. I would like to take the time to reproduce the issue.

**Comment by RobVerduijn** [comment]
The file listing from the ansible tower venv rpm  grepping all the lines with aiohttp

since that rpm is installed on the tower server the aiohttp python module is present

```
#rpm -ql ansible-tower-venv-tower-3.7.3-1.el8at.x86_64 | grep aiohttp
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info/INSTALLER
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info/LICENSE.txt
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info/METADATA
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info/RECORD
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info/WHEEL
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp-3.6.2.dist-info/top_level.txt
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__init__.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/__init__.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/abc.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/base_protocol.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/client.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/client_exceptions.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/client_proto.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/client_reqrep.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/client_ws.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/connector.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/cookiejar.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/formdata.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/frozenlist.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/hdrs.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/helpers.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/http.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/http_exceptions.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/http_parser.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/http_websocket.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/http_writer.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/locks.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/log.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/multipart.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/payload.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/payload_streamer.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/pytest_plugin.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/resolver.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/signals.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/streams.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/tcp_helpers.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/test_utils.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/tracing.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/typedefs.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_app.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_exceptions.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_fileresponse.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_log.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_middlewares.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_protocol.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_request.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_response.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_routedef.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_runner.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_server.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_urldispatcher.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/web_ws.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/__pycache__/worker.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_cparser.pxd
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_find_header.c
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_find_header.h
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_find_header.pxd
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_frozenlist.c
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_frozenlist.pyx
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_headers.pxi
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_helpers.c
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_helpers.pyi
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_helpers.pyx
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_http_parser.c
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_http_parser.pyx
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_http_writer.c
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_http_writer.pyx
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_websocket.c
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/_websocket.pyx
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/abc.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/base_protocol.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/client.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/client_exceptions.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/client_proto.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/client_reqrep.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/client_ws.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/connector.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/cookiejar.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/formdata.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/frozenlist.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/frozenlist.pyi
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/hdrs.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/helpers.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/http.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/http_exceptions.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/http_parser.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/http_websocket.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/http_writer.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/locks.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/log.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/multipart.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/payload.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/payload_streamer.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/py.typed
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/pytest_plugin.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/resolver.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/signals.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/signals.pyi
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/streams.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/tcp_helpers.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/test_utils.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/tracing.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/typedefs.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_app.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_exceptions.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_fileresponse.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_log.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_middlewares.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_protocol.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_request.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_response.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_routedef.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_runner.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_server.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_urldispatcher.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/web_ws.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/aiohttp/worker.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/msrest/pipeline/__pycache__/aiohttp.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/msrest/pipeline/aiohttp.py
/var/lib/awx/venv/awx/lib/python3.6/site-packages/msrest/universal_http/__pycache__/aiohttp.cpython-36.pyc
/var/lib/awx/venv/awx/lib/python3.6/site-packages/msrest/universal_http/aiohttp.py
```

**Comment by goneri** [comment]
Interesting, `aiohttp` is indeed present, but in a python3.6 venv. So you can install `python3-aiohttp` in parallel. If you run RHEL, you can find the RPM on EPEL.

**Comment by RobVerduijn** [comment]
tower is a delicate product
adding the epel repository has some consequences.
Aiohttp has quite a few dependencoes it needs from epel as well some of them conflict with the packages from the tower repos.
And some epel packages have higher versions than the ones from the tower repos.

I am rather reluctant to replace the tower rpms with those from epel.

Unless your collection becomes an official ansible and tower endorsed and supported product, asking the tower support team with an official support subscription will result in a 'not supported' or 'out of scope' answer. ( I hear that a lot from them )

Rob

**Comment by goneri** [comment]
Yes. I totally understand your point. We are working on a solution for this problem that is called Execution Environment: https://docs.ansible.com/ansible-tower/latest/html/administration/external_execution_envs.html.
In the meantime, you can install the `python3-aiohttp` from EPEL, it should now conflict with Tower.

**Comment by mariolenz** [comment]
@goneri 
> In the meantime, you can install the `python3-aiohttp` from EPEL, it should now conflict with Tower.

I think you meant that it should no**t** conflict with Tower, otherwise this doesn't make sense. Why advise people to install a package that **does** conflict with Tower?

**Comment by vvaldez** [comment]
I can confirm this happens on Tower 3.8.1 on RHEL 8, but not RHEL 7. I installed `aiohttp` in the awx/ansible venv, and also setup a custom one and configured Tower to use it - in both cases `aiohttp` could not be found. After installing `python3-aiohttp` it worked.

**Comment by goneri** [comment]
Hi,

The recent Ansible Automation Platform 2.0 comes with a new feature called Execution Environment. The modules are called from a container that embeds the dependencies. aiohttp is included in this container. So you should not face this issue anymore.
